Dart/src/com/jetbrains/lang/dart/ide/formatter/DartPreFormatProcessor.java

```diff
@@ -0,0 +1,133 @@
+package com.jetbrains.lang.dart.ide.formatter;
+
+import com.intellij.lang.ASTNode;
+import com.intellij.openapi.editor.Document;
+import com.intellij.openapi.editor.RangeMarker;
+import com.intellij.openapi.module.Module;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.roots.ContentEntry;
+import com.intellij.openapi.roots.ModuleRootManager;
+import com.intellij.openapi.util.Key;
+import com.intellij.openapi.util.TextRange;
+import com.intellij.openapi.util.io.FileUtil;
+import com.intellij.openapi.vfs.VfsUtilCore;
+import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.psi.FileViewProvider;
+import com.intellij.psi.PsiDocumentManager;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiFile;
+import com.intellij.psi.codeStyle.CodeStyleSettings;
+import com.intellij.psi.codeStyle.CodeStyleSettingsManager;
+import com.intellij.psi.impl.source.codeStyle.PreFormatProcessor;
+import com.jetbrains.lang.dart.DartLanguage;
+import com.jetbrains.lang.dart.analyzer.DartAnalysisServerService;
+import com.jetbrains.lang.dart.sdk.DartSdk;
+import com.jetbrains.lang.dart.sdk.DartSdkGlobalLibUtil;
+import org.dartlang.analysis.server.protocol.SourceEdit;
+import org.jetbrains.annotations.NotNull;
+
+import java.io.*;
+import java.util.Collection;
+import java.util.List;
+
+public class DartPreFormatProcessor implements PreFormatProcessor {
+  public static Key<Boolean> FORMAT_MARK = new Key<Boolean>("FORMAT_MARK");
+  public static Boolean FORMAT_MARKER = Boolean.TRUE;
+
+  private CodeStyleSettings mySettings;
```

Comment:
Yep, I see the test failing, and also can reproduce the same in runtime. Sample code 
```
main() {
  Iterable b = doSomething(2+2==4, 1+1, 2*2.0, "", [1], {1:2}, #*, null);
}
```
Quick fix starts a live template, template asks to reformat range, after DartPreFormatProcessor range for template variable becomes broken which seems to cause the mangled code.
![image](https://cloud.githubusercontent.com/assets/4499789/9553481/0a1eef86-4dc8-11e5-9875-c68b7261aeb5.png)
I'm afraid nobody can answer why that happens without some time (may be hours) of debugging. This might be a bug in template engine, template misuse, bug in PreformatProcessor or anything else. As an experiment we can try to use PostFormatProcessor instead of PreFormatProcessor.